### PR TITLE
Implemented BeamlineRayTracer as a set of free functions

### LINKED ISSUE #23005: Instrument Ray Tracer 2.0
`InstrumentRayTracer` works off `Instrument` not the `Instrument 2.0` layers, such as `ComponentInfo` [as demonstrated](https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/inc/MantidGeometry/Objects/InstrumentRayTracer.h#L56).  This is causing us to have a dependency on `Instrument` in `Peak` [here](https://github.com/mantidproject/mantid/blob/master/Framework/DataObjects/inc/MantidDataObjects/Peak.h#L182) and therefore we cannot get away from this legacy API. We are trying to stop developers needing to access `Instrument` as far as possible and use the new Geometry/Beamline layers instead.

Design a new "ray tracer" using `ComponentInfo` and consider how this could be rolled out. What the API would be like. What changes would have to be made to the code base? How will we ensure that performance is good enough? http://developer.mantidproject.org/DesignDocumentGuides.html?highlight=design

Create a branch then PR for the design. Put the design in here.
https://github.com/mantidproject/documents/tree/master/Design/Instrument-2.0

### PR DESCRIPTION
**Pull request by samueljackson92**
Created a new PR as the old one was 1) outdated, 2) had a messed up merge, 3) was based off of Bhuvan's repo (who has now left, but he still will get commit credit :) ). 

**Description of work.**
This PR introduces a new way to perform ray tracing. In order to move away from using `Instrument` 
and move towards using `Instrument 2.0` a new ray tracing system has been developed. The `BeamlineRayTracer` is composed of a set of free functions and uses `Beamline` layers such as `ComponentInfo` to perform ray tracing.  

**Report to:** @OwenArnold . 

**To test:**
Please test that the unit tests all pass.

Run the following command to run the unit tests:
`ctest -C Debug -R BeamlineRayTracer -V`

Fixes #23005 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [inline Framework/Geometry/test/BeamlineRayTracerTest.h]: I don't think this is necessary? If not remove "ConfigService" include and `using` too.

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [inline Framework/Geometry/test/BeamlineRayTracerTest.h]: It's not directly clear from the test setup that we expect it to apss through 2 components. Maybe we could add some description about that. i.e. what components are relevant from `create_instrument_and_componentInfo`

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [inline Framework/Geometry/test/BeamlineRayTracerTest.h]: I think this would be neater, and better future portable if done via componentInfo. i.e. with [indexOfAny](https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/inc/MantidGeometry/Instrument/ComponentInfo.h#L97) and [source](https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/inc/MantidGeometry/Instrument/ComponentInfo.h#L118) from ComponentInfo. Look thorough all test methods and try to avoid using old `IComponent` and derived where possible.

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [inline Framework/Geometry/test/BeamlineRayTracerTest.h]: Could also use `ComponentInfo::isDetector` as part of validation

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [inline Framework/Geometry/test/BeamlineRayTracerTest.h]: Same thing about avoiding IDs (see above)

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [inline Framework/Geometry/test/BeamlineRayTracerTest.h]: So the important note here is the the x offset is > the source radius? i.e. the source intersection is not counted.

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [inline Framework/Geometry/test/BeamlineRayTracerTest.h]: Method name and comment seem to be at odds. I thought only tubes were marked as outline components.

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [inline Framework/Geometry/test/BeamlineRayTracerTest.h]: Ok ignore comment above. This is where `ConfigService` is required.

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [inline Framework/Geometry/test/BeamlineRayTracerTest.h]: Do you want to include this block in the timings? Might be misleading for future maintainers.

File: Framework/Geometry/src/Objects/BeamlineRayTracer.cpp
Comment (OwenArnold) [inline Framework/Geometry/src/Objects/BeamlineRayTracer.cpp]: This is assigned but I think not used?

File: Framework/Geometry/src/Objects/BeamlineRayTracer.cpp
Comment (OwenArnold) [inline Framework/Geometry/src/Objects/BeamlineRayTracer.cpp]: It is a little weird to see it extracted like this. i.e treating a tube (effectively as line as far as centre positions are defined) like a quad. That's just an observation, the rest of it looks ok.

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Looks really nice. A few improvements to make.

I'm a bit worried that this immediately lives next to `InstrumentRayTracer`. Once merged is there a plan to deal with the technical debt that having both implementations will yield? Can we immediately roll this out and replace InstrumentRayTracer? If not would it be wiser to merge this into a long-running feature branch instead?

**Comment by samueljackson92** [comment]
@OwenArnold Thanks for the review. I'll work through the comments when I have time. 

About the roll out, long running feature branch sounds good 👍. So you're suggesting doing the roll out on another branch based off this one?

**Comment by OwenArnold** [comment]
@samueljackson92 I was thinking that you could create a feature branch off `master` called something like `ray-tracing-project` to which we could merge this.

I'm happy to debate this approach though. If you can guarantee a free week (or more) to make the rest of the changes to now eliminate `InstrumentRayTracer` i would be happy for this to go into master now.

**Comment by samueljackson92** [comment]
>  If you can guarantee a free week (or more) to make the rest of the changes 

I wish I do that 😞.

>  I was thinking that you could create a feature branch off master called something like ray-tracing-project to which we could merge this.

This sounds most sensible. I shall do this.

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (samueljackson92) [reply Framework/Geometry/test/BeamlineRayTracerTest.h]: How do I get from a component ID to an index? there doesn't seem to be an obvious option in compinfo?

Comment (OwenArnold) [reply Framework/Geometry/test/BeamlineRayTracerTest.h]: https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/inc/MantidGeometry/Instrument/ComponentInfo.h#L96

Comment (samueljackson92) [reply Framework/Geometry/test/BeamlineRayTracerTest.h]: But that's from a pointer to a component? not a component **id**?

Comment (samueljackson92) [reply Framework/Geometry/test/BeamlineRayTracerTest.h]: Or is _component ID == component index_ in all cases?

Comment (OwenArnold) [reply Framework/Geometry/test/BeamlineRayTracerTest.h]: It's the same thing. `Component::getID()` returns it's address.

Comment (OwenArnold) [reply Framework/Geometry/test/BeamlineRayTracerTest.h]: https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/src/Instrument/Component.cpp#L101 or more accurately https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/inc/MantidGeometry/IComponent.h#L34

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (samueljackson92) [reply Framework/Geometry/test/BeamlineRayTracerTest.h]: I think so, Bhuvan wrote this bit originally.

**Comment by samueljackson92** [comment]
@OwenArnold I've corrected for your comment. When this is passing the build I think it is in a state where it's worth creating the long running feature branch to start rolling it out.

Would you kindly take a quick final look?

File: Framework/Geometry/inc/MantidGeometry/Objects/BeamlineRayTracer.h
Comment (OwenArnold) [inline Framework/Geometry/inc/MantidGeometry/Objects/BeamlineRayTracer.h]: Why is this header-only? 🤔

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [inline Framework/Geometry/test/BeamlineRayTracerTest.h]: I think this could just become `componentInfo.type()` if you want a type check

instead of `bank->getXYForDetectorID(det.getID());` can just get the position directly and look at x,y. `componentInfo.position(detectorIndex)`.

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [inline Framework/Geometry/test/BeamlineRayTracerTest.h]: Same here as above. No need to use Instrument 1.

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (samueljackson92) [reply Framework/Geometry/test/BeamlineRayTracerTest.h]: There's a difference here. `getXYForDetectorID` is check the row/column index of the rectangular detector not the position. Perhaps a better name is `getRowColForDetector`?

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Minor changes. See comments above. Otherwise looking good.

File: Framework/Geometry/test/BeamlineRayTracerTest.h
Comment (OwenArnold) [reply Framework/Geometry/test/BeamlineRayTracerTest.h]: Ok, I see. I would still rather this was done without `Instrument 1.0` I think we can establish column  and row index using the `children` and other related functions on `ComponentInfo`. This could go in to a helper function in the test, or better yet into https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/inc/MantidGeometry/Instrument/ComponentInfoBankHelpers.h

Comment (samueljackson92) [reply Framework/Geometry/test/BeamlineRayTracerTest.h]: 👍 Agreed.

File: Framework/Geometry/inc/MantidGeometry/Objects/BeamlineRayTracer.h
Comment (OwenArnold) [reply Framework/Geometry/inc/MantidGeometry/Objects/BeamlineRayTracer.h]: I think that might be a hangover from how Bhuvan started to write this. I can't see why we don't have a separate implementation file, and I think I discussed that with him.

**Comment by samueljackson92** [comment]
@OwenArnold I believe this test failure is unrelated. This should be good to go now.

**Review by OwenArnold — DISMISSED** [review]

**Comment by peterfpeterson** [comment]
@samueljackson92 The windows build has a compiler warning that you need to fix.

**Comment by NickDraper** [comment]
@OwenArnold Could you take a look at this one. 
@martyngigg Could you take over the developer role

**Comment by martyngigg** [comment]
@OwenArnold Are you okay with this being pushed back to 4.1?

**Comment by NickDraper** [comment]
@martyngigg Another one to evaluate the value of.

**Comment by peterfpeterson** [comment]
This looks extremely useful. I think @OwenArnold needs to either approve it, or find somebody to finish it.

**Comment by AnthonyLim23** [comment]
@NickDraper this seems very stale, is it worth keeping or getting someone to redo it based from a recent master?

**Comment by NickDraper** [comment]
This branch is too stale to be useful now.  Closing this PR

**Pull request closed without merging**
